internal/api/impl/proxy/proxy.go

```diff
@@ -28,67 +29,181 @@ import (
 	"github.com/perses/perses/internal/api/interface/v1/globaldatasource"
 	"github.com/perses/perses/internal/api/interface/v1/globalsecret"
 	"github.com/perses/perses/internal/api/interface/v1/secret"
+	"github.com/perses/perses/internal/api/shared"
 	"github.com/perses/perses/internal/api/shared/crypto"
 	databaseModel "github.com/perses/perses/internal/api/shared/database/model"
+	"github.com/perses/perses/internal/api/shared/rbac"
 	"github.com/perses/perses/internal/api/shared/route"
 	"github.com/perses/perses/internal/api/shared/utils"
 	v1 "github.com/perses/perses/pkg/model/api/v1"
 	datasourceHTTP "github.com/perses/perses/pkg/model/api/v1/datasource/http"
+	"github.com/perses/perses/pkg/model/api/v1/role"
 	promConfig "github.com/prometheus/common/config"
 	"github.com/sirupsen/logrus"
 )
 
+// unsavedProxyBody is the body of the request when the datasource is not saved yet.
+// It contains the body of the request and the datasource spec, which is used to build the proxy rather than
+// retrieving the datasource from the database.
+type unsavedProxyBody struct {
+	Method string             `json:"method" yaml:"method"`
+	Body   []byte             `json:"body" yaml:"body"`
+	Spec   *v1.DatasourceSpec `json:"spec" yaml:"spec"`
```

Comment:
excepting the `body` attribute, all others attributes are mandatory. I suggest to :
- add the key `omitempty` for the body
- Override the method `UnmarshalJSON` to include the verification that `method` is not empty
- I guess we also need to check that `method` is a correct HTTP method
- And why not removing the pointer for `Spec`. Like that we are ensuring the `Spec` cannot be empty

Reply:
Done! I was tossing up doing validation in the `UnmarshalJSON`, or having an external `Validate` that was called after each `Bind`. Calling it in `Bind` would allow us to validate, regardless of Content-Type WDYT?

Reply:
Nah it's ok to do it in the `UnmarshalJSON` as there is no plan to support anything else than JSON when contacting the Perses HTTP API